### ISSUE TITLE
Runtime Public Key signature is not validated

### ISSUE THREAD

**Issue by CedarMist** [oasisprotocol/sapphire-paratime#135]
The public key is returned from a call to `oasis_callDataPublicKey` to oasis-web3-gateway.

https://github.com/oasisprotocol/sapphire-paratime/blob/2ee5b3868f56134e9b6ec2cdb2a87cd0104de0ce/clients/js/src/compat.ts#L506

Example response:

```python
{
  'jsonrpc': '2.0',
  'id': 3892329,
  'result': {
    'key': '0xd46745ab3a8a8467cc97d57cdcdcdff9ad5a628a4fd4898dfe5c3db331fc1050',
    'checksum': '0x59dd7e7188b9736897ca719ece9dce6d9618e3f487553dccbc601ba848b65d9c',
    'signature': '0x99a5f80cdc6df788495ba36b0d652901d49869f7bb2ff94db83c2a4379f5e4d39c7e9c6ed21ef2e379fa882cfa490f4139af33e01e554280a30d7b7ef2f74d02'
  }
}
```

However, the signature by the Key Manager provider ( `Sign(sk, key||checksum)` ) which authenticates the validity of the public key is not checked. 

This allows for the oasis-web3-gateway to MITM encrypted transaction data, which is bad!

### Suggested Fix

The public key of the Key Manager must be known to the client library via a route independent of the web3-gateway, or can be provided by the web3-gateway if there is some kind of signature by a long-term trusted keypair which can validate its authenticity.

Then the client must validate the signature before use, and must not submit encrypted transactions if validation fails.

This can be disabled for local testing nodes, but should be enabled on Testnet and *must* be enabled on Mainnet.

**Comment by nhynes** [comment]
> and must be enabled on Mainnet.

I agree with the sentiment, but it's important to remember that different people have different trust assumptions. For example, it is perfectly fine to to have client-side key verification if one trusts their gateway and its connection to its key manager.

That being said, thank you for making an issue about this, and I welcome a PR if you have one available!

**Comment by CedarMist** [comment]
Where do I find and validate the Key Manager public key, so I can validate the signature provided by oasis-web3-gateway?

**Comment by nhynes** [comment]
Thank you for taking this up! I'd bet that @pro-wh knows where you can find the KM public key. I checked the CLI and ts-web SDKs for getters, but didn't find any, so it would benefit from an expert's help here.

**Comment by kostko** [comment]
To fully verify this on the client, the client would require a consensus layer light client. In the upcoming version of Oasis Core, the KM public key is available in the [key manager status structure](https://github.com/oasisprotocol/oasis-core/blob/23d959134b95d28644b995ac84063467abd22ea1/go/keymanager/api/api.go#L132) that is published in the consensus layer.

Otherwise it needs access to a trusted node to read consensus state (in which case it may be fine if it also trusts the web3 gateway).
